### ISSUE TITLE
Gluon dense layer gets segfault on large arrays when MKLDNN is built in MXNet.

### ISSUE THREAD

**Issue by zheng-da** [apache/incubator-mxnet#13451]
The code below gets segfault when MKLDNN is built in MXNet.

```python
import mxnet as mx
from mxnet import gluon
def run_linear():
    data = mx.nd.ones(shape=(50*1000*1000, 100))
    linear = gluon.nn.Dense(100)
    linear.initialize(ctx=mx.cpu(0))
    res = linear(data)
    res.wait_to_read()

run_linear()
```

**Comment by pengzhao-intel** [comment]
Thanks for the case. MKL-DNN has the plan to support large inputs and will back you later.

**Comment by TaoLv** [comment]
@mxnet-label-bot add [MKLDNN, Gluon]

**Comment by TaoLv** [comment]
Currently MKL-DNN doesn't support large dimension (> INT_MAX) and that will be change in the v1.0 release of it. Before that, we need fallback the computation to MXNet if the input/output dimension is larger than INT_MAX.
@zheng-da Do you know if large NDArray is already enabled in MXNet or it is just an effort ongoing? Do we need to change MKL-DNN backend in the coming release?

**Comment by zheng-da** [comment]
@pengzhao-intel @TaoLv This is a critical bug for the DGL release. @apeforest is working on fixing all mxnet operators. could you please fix this one as well. Thanks.

**Comment by TaoLv** [comment]
I‘m working on this issue. Actually, `50*1000*1000` doesn't exceed the range of int32. This case works well with `use_bias=False` on my side.

**Comment by zheng-da** [comment]
@TaoLv a matrix of 50*1000*1000 rows and 100 columns has 5B elements. It exceeds the range of int32.

**Comment by TaoLv** [comment]
Yes, but I think that should be covered by MKL-DNN. If not, it's absolutely a bug. I'm working with MKL-DNN team to root cause this issue. Will update to you if any progress. Thanks.

**Comment by TaoLv** [comment]
Closed via #13668 .
